src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -143,151 +131,77 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
-        /// </summary>
-        private async ValueTask SetLastRequestedSolutionAsync(
+        private async ValueTask<ReferenceCountedDisposable<LazySolution>> GetOrCreateSolutionAsync(
+            AssetProvider assetProvider,
             Checksum solutionChecksum,
-            bool fromPrimaryBranch,
-            ReferenceCountedDisposable<LazySolution> solution,
+            int workspaceVersion,
+            bool updatePrimaryBranch,
             CancellationToken cancellationToken)
         {
-            // Collect any existing solutions we're pointing at so we can decrease their ref count.  Do this out of
-            // the lock to prevent deadlocks.
-            using var solutionsToDispose = TemporaryArray<ReferenceCountedDisposable<LazySolution>>.Empty;
-
-            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
-            {
-                // All the TryAddReference calls below must succeed as we're called from a caller that is pinning
-                // solution anyways with it's own refcount of at least 1.
-
-                solutionsToDispose.Add(_lastRequestedAnyBranchSolution.refCountedSolution);
-                _lastRequestedAnyBranchSolution = (solutionChecksum, solution.TryAddReference() ?? throw ExceptionUtilities.Unreachable);
-
-                if (fromPrimaryBranch)
-                {
-                    solutionsToDispose.Add(_lastRequestedPrimaryBranchSolution.refCountedSolution);
-                    _lastRequestedPrimaryBranchSolution = (solutionChecksum, solution.TryAddReference() ?? throw ExceptionUtilities.Unreachable);
-                }
-            }
-
-            foreach (var solutionToDispose in solutionsToDispose)
+            // Always try to retrieve cached solutions from the primary branch first.  That way we can use the solutions
+            // that were the real solutions of this Workspace, and not ones forked off from that.  This gives the
+            // highest likelihood of sharing data and being able to reuse caches and services shared among all
+            // components.
+            var primaryBranchRefCountedSolution = await _primaryBranchSolutionCache.TryFastGetSolutionAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+            if (primaryBranchRefCountedSolution != null)
+                return primaryBranchRefCountedSolution;
+
+            // Otherwise, have the any-branch solution try to retrieve or create the solution.
+            var anyBranchRefCountedSolution =
+                await _anyBranchSolutionCache.TryFastGetSolutionAsync(solutionChecksum, cancellationToken).ConfigureAwait(false) ??
+                await _anyBranchSolutionCache.SlowGetOrCreateSolutionAsync(
+                    solutionChecksum,
+                    cancellationToken => ComputeSolutionAsync(assetProvider, solutionChecksum, cancellationToken),
+                    cancellationToken).ConfigureAwait(false);
```

Comment:
note: teh lambda we pass in here is the same code that used to be in SlowGetSolutionAsync at this line:

https://github.com/dotnet/roslyn/pull/62872/files#diff-fedbb1bbc0cee7c7ea4dbd51b8b27023e351ec9a062fb4623fdb8063b5517d30L271